sh-utils-1.13/man/id.1
.TH ID 1 "GNU Shell Utilities" "FSF" \" -*- nroff -*-
.SH NAME
id \- print real and effective UIDs and GIDs
.SH SYNOPSIS
.B id
[\-gnruG] [\-\-group] [\-\-name] [\-\-real] [\-\-user] [\-\-groups]
[\-\-help] [\-\-version] [username]
.SH DESCRIPTION
This documentation is no longer being maintained and may be inaccurate
or incomplete.  The Texinfo documentation is now the authoritative source.
.PP
This manual page
documents the GNU version of
.BR id .
.B id
prints information about the given user, or the process running it if
no user is specified.  By default, it prints the real user ID, real
group ID, effective user ID if different from the real user ID,
effective group ID if different from the real group ID, and
supplemental group IDs.  It prints each of these values preceded by a
string identifying it and followed by the corresponding user or group
name in parentheses.
.PP
The options cause
.B id
to print only part of the above information.
.SS OPTIONS
.TP
.I "\-g, \-\-group"
Print only the group ID.
.TP
.I "\-G, \-\-groups"
Print only the supplementary groups.
.TP
.I "\-\-help"
Print a usage message on standard output and exit successfully.
.TP
.I "\-n, \-\-name"
Print the user or group name instead of the ID number.
Requires
.IR \-u ,
.IR \-g ,
or
.IR \-G .
.TP
.I "\-r, \-\-real"
Print the real, instead of effective, user or group ID.
Requires
.IR \-u ,
.IR \-g ,
or
.IR \-G .
.TP
.I "\-u, \-\-user"
Print only the user ID.
.TP
.I "\-\-version"
Print version information on standard output then exit successfully.


m4-1.4/NEWS
GNU m4 NEWS - User visible changes.
Copyright (C) 1992, 1993, 1994 Free Software Foundation, Inc.

Version 1.4 - October 1994, by Franc,ois Pinard

(No user visible changes)

Version 1.3 - September 1994, by Franc,ois Pinard

* Diversions are created as needed.  Option `-N' is still accepted, but
otherwise ignored.  Users should use only negative diversion numbers,
instead of high positive numbers, for diverting to nowhere.

* Diversions should also work faster.  No temporary files will be needed
at all if all diversions taken altogether do not use more than 512K.

* Frozen state files may be produced with the `--freeze-state' (-F)
option and later brought back through the `--reload-state' (-R) option.

Version 1.2 - July 1994, by Franc,ois Pinard

* In patsubst(STRING, REGEXP, REPLACEMENT), \& in REPLACEMENT has been
changed to represent this part of STRING matched by the whole REGEXP,
instead of the whole STRING as before.  \0 does the same, but emits a
diagnostic saying it will disappear in some subsequent release.

* eval(EXPR) emits a diagnostic if EXPR has suffixed crumb.  The same for
other numeric conversions in incr(), decr(), divert(), etc.

* `--fatal-warnings' (-E) stops execution at first warning.

* `--nesting-limit=LEVEL' (-L LEVEL) sets a limit to macro nesting.
It is initially fixed at 250.

* `--word-regexp=REGEXP' (-W REGEXP) modifies macro name syntax, like
does the new `changeword(REGEXP)' macro.  This feature is experimental,
tell me your opinions about it.  You do need --enable-changeword at
configure time to get these things.  Do *not* depend on them yet.

* Trace output format is scannable by GNU Emacs' next-error function.

* Stack overflow is detected and diagnosed on some capable systems.

* Various bugs have been corrected, m4 should be more portable.  See the
ChangeLog for details.

Version 1.1 - November 1993, by Franc,ois Pinard

Changes which might affect existing GNU m4 scripts:

* Option `-V' has been removed, use `--version' instead.  `--version'
writes on standard output instead of standard error, and inhibits any
script execution.

* `--no-gnu-extensions' has been renamed `--traditional'.

* In `eval', `^' used to indicate exponentiation, use `**' instead.

* The automatic undiversion which takes place at end of all input is
forced into the main output stream.

Changes which are unlikely to affect existing scripts:

* `--help' prints an usage summary on standard output.  Script execution
is then inhibited.

* `--prefix-builtins' (-P) prefixes all builtin macros by `m4_'.

* Most builtin macros for which arguments are mandatory, called without
any arguments, are no more recognized as builtin macros: they are
consequently copied verbatim to the output stream.

* `define' and `pushdef' are usable with only one argument, they give
this argument an empty definition.

* `eval' new operators for binary representation handling: `^' for
exclusive-or, `~' for the bitwise negation, `<<' and `>>' for shifts.

* `eval' recognizes the notation 0bDIGITS for binary numbers and the
notation 0rRADIX:DIGITS for numbers in any radix from 1 to 36.

Version 1.0.3 - December 1992, by Franc,ois Pinard

Changes for the user:

* `dnl' outputs a diagnostic if immediately followed by `('.  Usually,
`dnl' is followed by newline or whitespace.

* `ifelse' accepts without complaining the common idiom of having only
one argument.  This is useful for introducing long comments.

* `eval' always expresses values as signed, whatever the radix.

* M4OPTS environment variable is no longer obeyed.

* `--no-warnings' option is renamed `--silent'.

* Debug lines use a new format more compatible with GNU standards.

* Various bugs have been corrected.  See the ChangeLog for details.

Changes for the installer:

* GNU m4 now uses an Autoconf-generated configure script, and should be
more easily portable in many ways.  (Cray is not supported yet).

* `make check' has been made more portable, expect no errors.

Changes for the programmer:

* Sources have been fully reindented to comply with GNU standards, and
cleaned up in many ways.

* Sources have been protoized.  Non-ANSI compilers are automatically
detected, then sources are unprotoized on the fly before compilation.

* GNU m4 uses newer versions of obstack, regex, getopt, etc.

Version 1.0 - October 1991, by Rene' Seindal

* Uses GNU configure, taken from the gdb distribution.

* Uses GNU getopt(), with long option names.

* The -Q/+quiet option is added, which suppresses warnings about missing
or superflous arguments to built-in macros.

* Added default options via the M4OPTS environment variable.

* Several minor bugs have been fixed.

Version 0.99 - July 1991, by Rene' Seindal

* The builtins `incr' and `decr' are now implemented without use of
`eval'.

* The builtin `indir' is added, to allow for indirect macro calls
(allows use of "illegal" macro names).

* The debugging and tracing facilities has been enhanced considerably.
See the manual for details.

* The -tMACRO option is added, marks MACRO for tracing as soon as it
is defined.

* Builtins are traced after renaming iff they were before.

* Named files can now be undiverted.

* The -Nnum option can be used to increase the number of divertions
available.

* Calling changecom without arguments now disables all comment handling.

* A bug in `dnl' is fixed.

* A bug in the multi-character quoting code is fixed.

* Several typos in the manual has been corrected.  More probably persist.

Version 0.75 - November 1990, by Rene' Seindal

* Implemented search path for include files (-I option and M4PATH
environment variable).

* Implemented builtin `format' for printf-like formatting.

* Implemented builtin `regexp' for searching for regular expressions.


shellutils-1.9.4/src/id.c
/* id -- print real and effective UIDs and GIDs
   Copyright (C) 89, 90, 91, 92, 93, 1994 Free Software Foundation, Inc.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.  */

/* Written by Arnold Robbins, arnold@audiofax.com.
   Major rewrite by David MacKenzie, djm@gnu.ai.mit.edu. */

#ifdef HAVE_CONFIG_H
#if defined (CONFIG_BROKETS)
/* We use <config.h> instead of "config.h" so that a compilation
   using -I. -I$srcdir will use ./config.h rather than $srcdir/config.h
   (which it would do because it found this file in $srcdir).  */
#include <config.h>
#else
#include "config.h"
#endif
#endif

#include <stdio.h>
#include <getopt.h>
#include <sys/types.h>
#include <pwd.h>
#include <grp.h>
#include <getopt.h>

#include "version.h"
#include "system.h"

#ifdef _POSIX_VERSION
#include <limits.h>
#if !defined(NGROUPS_MAX) || NGROUPS_MAX < 1
#undef NGROUPS_MAX
#define NGROUPS_MAX sysconf (_SC_NGROUPS_MAX)
#endif /* !NGROUPS_MAX */

#else /* not _POSIX_VERSION */
struct passwd *getpwuid ();
struct group *getgrgid ();
uid_t getuid ();
gid_t getgid ();
uid_t geteuid ();
gid_t getegid ();
#include <sys/param.h>
#if !defined(NGROUPS_MAX) && defined(NGROUPS)
#define NGROUPS_MAX NGROUPS
#endif /* not NGROUPS_MAX and NGROUPS */
#endif /* not _POSIX_VERSION */

char *xmalloc ();
int getugroups ();
void error ();

static void print_user ();
static void print_group ();
static void print_group_list ();
static void print_full_info ();
static void usage ();

/* The name this program was run with. */
char *program_name;

/* If nonzero, output only the group ID(s). -g */
static int just_group = 0;

/* If nonzero, output user/group name instead of ID number. -n */
static int use_name = 0;

/* If nonzero, output real UID/GID instead of default effective UID/GID. -r */
static int use_real = 0;

/* If nonzero, output only the user ID(s). -u */
static int just_user = 0;

/* If nonzero, output only the supplementary groups. -G */
static int just_group_list = 0;

/* The real and effective IDs of the user to print. */
static uid_t ruid, euid;
static gid_t rgid, egid;

/* The number of errors encountered so far. */
static int problems = 0;

/* If non-zero, display usage information and exit.  */
static int show_help;

/* If non-zero, print the version on standard output and exit.  */
static int show_version;

static struct option const longopts[] =
{
  {"group", no_argument, NULL, 'g'},
  {"groups", no_argument, NULL, 'G'},
  {"help", no_argument, &show_help, 1},
  {"name", no_argument, NULL, 'n'},
  {"real", no_argument, NULL, 'r'},
  {"user", no_argument, NULL, 'u'},
  {"version", no_argument, &show_version, 1},
  {NULL, 0, NULL, 0}
};

void
main (argc, argv)
     int argc;
     char **argv;
{
  int optc;

  program_name = argv[0];

  while ((optc = getopt_long (argc, argv, "gnruG", longopts, (int *) 0))
	 != EOF)
    {
      switch (optc)
	{
	case 0:
	  break;
	case 'g':
	  just_group = 1;
	  break;
	case 'n':
	  use_name = 1;
	  break;
	case 'r':
	  use_real = 1;
	  break;
	case 'u':
	  just_user = 1;
	  break;
	case 'G':
	  just_group_list = 1;
	  break;
	default:
	  usage (1);
	}
    }

  if (show_version)
    {
      printf ("%s\n", version_string);
      exit (0);
    }

  if (show_help)
    usage (0);

  if (just_user + just_group + just_group_list > 1)
    error (1, 0, "cannot print only user and only group");

  if (just_user + just_group + just_group_list == 0 && (use_real || use_name))
    error (1, 0, "cannot print only names or real IDs in default format");

  if (argc - optind > 1)
    usage (1);

  if (argc - optind == 1)
    {
      struct passwd *pwd = getpwnam (argv[optind]);
      if (pwd == NULL)
	error (1, 0, "%s: No such user", argv[optind]);
      ruid = euid = pwd->pw_uid;
      rgid = egid = pwd->pw_gid;
    }
  else
    {
      euid = geteuid ();
      ruid = getuid ();
      egid = getegid ();
      rgid = getgid ();
    }

  if (just_user)
    print_user (use_real ? ruid : euid);
  else if (just_group)
    print_group (use_real ? rgid : egid);
  else if (just_group_list)
    print_group_list (argv[optind]);
  else
    print_full_info (argv[optind]);
  putchar ('\n');

  exit (problems != 0);
}

/* Print the name or value of user ID UID. */

static void
print_user (uid)
     int uid;
{
  struct passwd *pwd = NULL;

  if (use_name)
    {
      pwd = getpwuid (uid);
      if (pwd == NULL)
	problems++;
    }

  if (pwd == NULL)
    printf ("%u", (unsigned) uid);
  else
    printf ("%s", pwd->pw_name);
}

/* Print the name or value of group ID GID. */

static void
print_group (gid)
     int gid;
{
  struct group *grp = NULL;

  if (use_name)
    {
      grp = getgrgid (gid);
      if (grp == NULL)
	problems++;
    }

  if (grp == NULL)
    printf ("%u", (unsigned) gid);
  else
    printf ("%s", grp->gr_name);
}

/* Print all of the distinct groups the user is in . */

static void
print_group_list (username)
     char *username;
{
  print_group (rgid);
  if (egid != rgid)
    {
      putchar (' ');
      print_group (egid);
    }

#ifdef NGROUPS_MAX
  {
    int ngroups;
    GETGROUPS_T *groups;
    register int i;

    groups = (GETGROUPS_T *) xmalloc (NGROUPS_MAX * sizeof (GETGROUPS_T));
    if (username == 0)
      ngroups = getgroups (NGROUPS_MAX, groups);
    else
      ngroups = getugroups (NGROUPS_MAX, groups, username);
    if (ngroups < 0)
      {
	error (0, errno, "cannot get supplemental group list");
	problems++;
	free (groups);
	return;
      }

    for (i = 0; i < ngroups; i++)
      if (groups[i] != rgid && groups[i] != egid)
	{
	  putchar (' ');
	  print_group (groups[i]);
	}
    free (groups);
  }
#endif
}

/* Print all of the info about the user's user and group IDs. */

static void
print_full_info (username)
     char *username;
{
  struct passwd *pwd;
  struct group *grp;

  printf ("uid=%u", (unsigned) ruid);
  pwd = getpwuid (ruid);
  if (pwd == NULL)
    problems++;
  else
    printf ("(%s)", pwd->pw_name);
  
  printf (" gid=%u", (unsigned) rgid);
  grp = getgrgid (rgid);
  if (grp == NULL)
    problems++;
  else
    printf ("(%s)", grp->gr_name);
  
  if (euid != ruid)
    {
      printf (" euid=%u", (unsigned) euid);
      pwd = getpwuid (euid);
      if (pwd == NULL)
	problems++;
      else
	printf ("(%s)", pwd->pw_name);
    }
  
  if (egid != rgid)
    {
      printf (" egid=%u", (unsigned) egid);
      grp = getgrgid (egid);
      if (grp == NULL)
	problems++;
      else
	printf ("(%s)", grp->gr_name);
    }

#ifdef NGROUPS_MAX
  {
    int ngroups;
    GETGROUPS_T *groups;
    register int i;

    groups = (GETGROUPS_T *) xmalloc (NGROUPS_MAX * sizeof (GETGROUPS_T));
    if (username == 0)
      ngroups = getgroups (NGROUPS_MAX, groups);
    else
      ngroups = getugroups (NGROUPS_MAX, groups, username);
    if (ngroups < 0)
      {
	error (0, errno, "cannot get supplemental group list");
	problems++;
	free (groups);
	return;
      }

    if (ngroups > 0)
      fputs (" groups=", stdout);
    for (i = 0; i < ngroups; i++)
      {
	if (i > 0)
	  putchar (',');
	printf ("%u", (unsigned) groups[i]);
	grp = getgrgid (groups[i]);
	if (grp == NULL)
	  problems++;
	else
	  printf ("(%s)", grp->gr_name);
      }
    free (groups);
  }
#endif
}

static void
usage (status)
     int status;
{
  if (status != 0)
    fprintf (stderr, "Try `%s --help' for more information.\n",
	     program_name);
  else
    {
      printf ("Usage: %s [OPTION]... [USERNAME]\n", program_name);
      printf ("\
\n\
  -g, --group     print only the group ID\n\
  -G, --groups    print only the supplementary groups\n\
  -n, --name      print a name instead of a number, for -ugG\n\
  -r, --real      print the real ID instead of effective ID, for -ugG\n\
  -u, --user      print only the user ID\n\
      --help      display this help and exit\n\
      --version   output version information and exit\n\
\n\
Without any OPTION, print some useful set of identified information.\n\
");
    }
  exit (status);
}
